### PR TITLE
[#4101] feat(core): Support PostgreSQL storage backend

### PR DESCRIPTION
**Pull request by yuqi1129**
### What changes were proposed in this pull request?

- Support PostgreSQL JDBC backend.

### Why are the changes needed?

User need.

Fix: #4101 

### Does this PR introduce _any_ user-facing change?

N/A

### How was this patch tested?

Existing ITs

### CODE REVIEW

File: core/src/main/java/org/apache/gravitino/storage/relational/mapper/CatalogMetaMapper.java [hunk 1]
```diff
@@ -38,148 +38,168 @@
...
+        databaseId = "postgresql"),
+  })
   List<CatalogPO> listCatalogPOsByMetalakeId(@Param("metalakeId") Long metalakeId);
 
-  @Select(
-      "SELECT catalog_id as catalogId FROM "
-          + TABLE_NAME
-          + " WHERE metalake_id = #{metalakeId} AND catalog_name = #{catalogName} AND deleted_at = 0")
+  @SelectProvider.List({
+    @SelectProvider(
```
Comment (yuqi1129) [inline core/src/main/java/org/apache/gravitino/storage/relational/mapper/CatalogMetaMapper.java:58]: This is method one.

File: core/src/main/java/org/apache/gravitino/storage/relational/mapper/MetalakeMetaSQLProvider.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+import org.apache.gravitino.storage.relational.po.MetalakePO;
+import org.apache.ibatis.annotations.Param;
+
+/** SQL Provider for Metalake Meta operations. */
+public class MetalakeMetaSQLProvider {
+
+  private final MetalakeMetaBaseProvider provider;
+
+  public MetalakeMetaSQLProvider() {
+    provider = MetalakeMetaBaseProvider.getProvider();
```
Comment (yuqi1129) [inline core/src/main/java/org/apache/gravitino/storage/relational/mapper/MetalakeMetaSQLProvider.java:31]: This is method two.

File: core/src/main/java/org/apache/gravitino/storage/relational/mapper/CatalogMetaPGProvider.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.gravitino.storage.relational.mapper;
+
+public class CatalogMetaPGProvider extends CatalogMetaMySQLProvider {}
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/storage/relational/mapper/CatalogMetaPGProvider.java:21]: Use Postgres, not PG.

**Comment by jerryshao** [comment]
I'm OK with the current solution, you can go ahead, @yuqi1129

**Comment by yuqi1129** [comment]
The related documents will be added to another PR, this one is already too large.

File: .github/workflows/backend-integration-test.yml [hunk 4]
```diff
@@ -132,11 +128,7 @@ jobs:
         architecture: [ linux/amd64 ]
         java-version: [ 17 ]
         test-mode: [ embedded, deploy ]
-        include:
-          - test-mode: 'embedded'
-            backend: 'h2'
-          - test-mode: 'deploy'
-            backend: 'mysql'
+        backend: [ h2, mysql, postgresql ]
```
Comment (yuqi1129) [inline .github/workflows/backend-integration-test.yml:131]: @mchades , I 've reverted your change here; no better solution has been found. Please take a look if you are free.

Comment (mchades) [reply .github/workflows/backend-integration-test.yml:131]: you can use `exclude`, FYI: https://docs.github.com/en/enterprise-cloud@latest/actions/writing-workflows/choosing-what-your-workflow-does/running-variations-of-jobs-in-a-workflow#excluding-matrix-configurations

Comment (yuqi1129) [reply .github/workflows/backend-integration-test.yml:131]: I already checked it and it's a bit complicated for use to configure accurately, let me think about it.

Comment (mchades) [reply .github/workflows/backend-integration-test.yml:131]: try this:
```
matrix:
        architecture: [linux/amd64]
        java-version: [ 17 ]
        test-mode: [embedded, deploy]
        backend: [h2, mysql, postgresql]
        exclude:
          - test-mode: 'embedded'
            backend: 'mysql'
          - test-mode: 'embedded'
            backend: 'postgresql'
          - test-mode: 'deploy'
            backend: 'h2'
```

**Comment by jerryshao** [comment]
Please split your PR into small ones, you have to figure out how to split the PRs, otherwise it is impossible to review.

File: .github/workflows/backend-integration-test.yml (hunk 4 above)
Comment (yuqi1129) [reply .github/workflows/backend-integration-test.yml:138]: Done.

File: catalogs/catalog-hadoop/build.gradle.kts [hunk 5]
```diff
@@ -79,6 +79,7 @@ dependencies {
   testImplementation(libs.mockito.core)
   testImplementation(libs.mockito.inline)
   testImplementation(libs.mysql.driver)
+  testImplementation(libs.postgresql.driver)
```
Comment (jerryshao) [inline catalogs/catalog-hadoop/build.gradle.kts:82]: Do we need to install pg driver manually when we launch Gravitino?

Comment (jerryshao) [reply catalogs/catalog-hadoop/build.gradle.kts:82]: I think we need to update the doc accordingly, to let user understand how to use PG backend.

File: core/src/main/java/org/apache/gravitino/storage/relational/mapper/CatalogMetaSQLProviderFactory.java [hunk 6]
```diff
@@ -49,6 +52,88 @@ static class CatalogMetaMySQLProvider extends CatalogMetaBaseSQLProvider {}
 
   static class CatalogMetaH2Provider extends CatalogMetaBaseSQLProvider {}
 
+  static class CatalogMetaPostgreSQLProvider extends CatalogMetaBaseSQLProvider {
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/storage/relational/mapper/CatalogMetaSQLProviderFactory.java:55]: Can we have a package "postgresql" under this and move all the pg class to that folder.

Besides, we have several changes for PG, but we don't have UTs to cover the pg support, only have IT. I think it is not enough, we should add UTs for PG, as well as for MySQL.

File: scripts/postgresql/schema-0.7.0-postgresql.sql [hunk 7]
```diff
@@ -0,0 +1,450 @@
...
+--
+-- Unless required by applicable law or agreed to in writing,
+-- software distributed under the License is distributed on an
+-- "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+-- KIND, either express or implied.  See the License for the
+-- specific language governing permissions and limitations
+-- under the License.
+--
+
+CREATE TABLE IF NOT EXISTS metalake_meta (
```
[the remark this answers is not in the document]
Comment (jerryshao) [reply scripts/postgresql/schema-0.7.0-postgresql.sql:20]: Where do we specify the db and schema for PG?

File: catalogs/catalog-hadoop/build.gradle.kts (hunk 5 above)
Comment (yuqi1129) [reply catalogs/catalog-hadoop/build.gradle.kts:82]: > Do we need to install pg driver manually when we launch Gravitino?

Yes, it's similar to MySQL backend and we need to download the corresponding JDBC driver to the classpath.

Comment (yuqi1129) [reply catalogs/catalog-hadoop/build.gradle.kts:82]: > I think we need to update the doc accordingly, to let user understand how to use PG backend.

I will use another PR to refine the document about it.

File: core/src/main/java/org/apache/gravitino/storage/relational/mapper/CatalogMetaSQLProviderFactory.java (hunk 6 above)
Comment (yuqi1129) [reply core/src/main/java/org/apache/gravitino/storage/relational/mapper/CatalogMetaSQLProviderFactory.java:55]: > Can we have a package "postgresql" under this and move all the pg class to that folder.

Okay.

> Besides, we have several changes for PG, but we don't have UTs to cover the pg support, only have IT. I think it is not enough, we should add UTs for PG, as well as for MySQL.

Yeah, I will try to boost a separate PR to test the MySQL and PG backend in UT.

File: scripts/postgresql/schema-0.7.0-postgresql.sql (hunk 7 above)
[the remark this answers is not in the document]
Comment (yuqi1129) [reply scripts/postgresql/schema-0.7.0-postgresql.sql:20]: > What is the supported version for PG?

I only test PG12 and PG13, let me verify more versions and add some description in the create schema script. 

> Where do we specify the db and schema for PG?

It's similar to MySQL backend. We need to specify it in the JDBC URL like `jdbc:postgresql://192.168.215.4:5432/PG_JDBC_BACKEND?currentSchema=pTLwBCewFm`.

File: core/src/main/java/org/apache/gravitino/storage/relational/mapper/postgresql/FilesetMetaPostgreSQLProvider.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+import org.apache.gravitino.storage.relational.mapper.FilesetMetaBaseSQLProvider;
+import org.apache.gravitino.storage.relational.po.FilesetPO;
+
+public class FilesetMetaPostgreSQLProvider extends FilesetMetaBaseSQLProvider {
+  @Override
+  public String softDeleteFilesetMetasByMetalakeId(Long metalakeId) {
+    return "UPDATE "
+        + META_TABLE_NAME
+        + " SET deleted_at = floor(extract(epoch from((current_timestamp - timestamp '1970-01-01 00:00:00')*1000)))"
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/storage/relational/mapper/postgresql/FilesetMetaPostgreSQLProvider.java:31]: I guess this line is longer than 100 chars, can you please fix all of such problems?

Comment (yuqi1129) [reply core/src/main/java/org/apache/gravitino/storage/relational/mapper/postgresql/FilesetMetaPostgreSQLProvider.java]: done.

**Review by jerryshao — APPROVED** [review]

**Pull request merged**